rafthttp/stream.go

```diff
@@ -392,45 +392,63 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
 	lv := semver.Must(semver.NewVersion(version.Version))
 	if compareMajorMinorVersion(rv, lv) == -1 && !checkStreamSupport(rv, t) {
 		resp.Body.Close()
+		cr.picker.unreachable(u)
 		return nil, errUnsupportedStreamType
 	}
 
-	switch resp.StatusCode {
-	case http.StatusGone:
-		resp.Body.Close()
-		err := fmt.Errorf("the member has been permanently removed from the cluster")
-		select {
-		case cr.errorc <- err:
-		default:
+	err = cr.checkResponseStatusCode(resp)
+	if err != nil {
+		cr.picker.unreachable(u)
+		// errMemberRemoved is a critical error since a removed member should
+		// always be stopped. So we use reportCriticalError to report it to errorc.
+		if err == errMemberRemoved {
+			reportCriticalError(err, cr.errorc)
 		}
 		return nil, err
-	case http.StatusOK:
-		return resp.Body, nil
+	}
+
+	return resp.Body, nil
+}
+
+// checkResponseStatusCode checks the response of one stream request.
+// It returns nil if the response indicates that stream request succeeds,
+// or it returns error.
+// If there is an error, the function always reads the response body
+// and closes it.
+func (cr *streamReader) checkResponseStatusCode(resp *http.Response) error {
+	if resp.StatusCode == http.StatusOK {
+		return nil
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
```

Comment:
> The HTTP status that returns by stream and post are different in some cases. 

You are wrong. They have exact same logic to handle `http.StatusPreconditionFailed` and `http.StatusForbidden`. Please do not repeat them here. Also please make the names of the two checking function to be consistent and put them into the consistent file. (not one in util and one in the pipeline)